chain/chain/src/resharding/resharding_actor.rs

```diff
@@ -1,111 +1,196 @@
+                        act.schedule_resharding(parent_shard_uid, ctx);
                     },
                 );
             }
+            // The event is already started, no need to reschedule.
+            ReshardingSchedulingStatus::AlreadyStarted => {}
         }
     }
 
-    fn handle_flat_storage_catchup(
+    // function to check if any one of the resharding block candidates is final
+    // and part of the canonical chain.
+    fn get_resharding_scheduling_status(
         &self,
-        resharder: FlatStorageResharder,
-        shard_uid: ShardUId,
-        ctx: &mut dyn DelayedActionRunner<Self>,
-    ) {
-        match resharder.shard_catchup_task(shard_uid, &self.chain_store) {
-            FlatStorageReshardingTaskResult::Successful { .. } => {
-                // All good.
-            }
-            FlatStorageReshardingTaskResult::Failed => {
-                panic!("impossible to recover from a flat storage shard catchup failure!")
-            }
-            FlatStorageReshardingTaskResult::Cancelled => {
-                // The task has been cancelled. Nothing else to do.
+        parent_shard_uid: ShardUId,
+    ) -> ReshardingSchedulingStatus {
+        tracing::info!(target: "resharding", ?parent_shard_uid, "get_resharding_scheduling_status");
+
+        if self.resharding_started.contains(&parent_shard_uid) {
+            // The event is already in progress, no need to reschedule.
+            tracing::info!(target: "resharding", "resharding already in progress");
+            return ReshardingSchedulingStatus::AlreadyStarted;
+        }
+
+        let events = self.resharding_events.get(&parent_shard_uid).unwrap();
+
+        let chain_final_height = self.chain_store.final_head().unwrap().height;
+        for event in events {
+            tracing::info!(
+                "get_resharding_scheduling_status: head height: {}, resharding_block: {:?}",
+                chain_final_height,
+                event.resharding_block,
+            );
+
+            // To check whether we can start resharding, we need to check if the resharding block is final.
+            // We check if the resharding block is behind the final block and is part of the canonical chain.
+            if event.resharding_block.height > chain_final_height {
+                continue;
             }
-            FlatStorageReshardingTaskResult::Postponed => {
-                // The task must be retried later.
-                ctx.run_later(
-                    "ReshardingActor FlatStorageCatchup",
-                    Duration::milliseconds(1000),
-                    move |act, ctx| {
-                        act.handle_flat_storage_catchup(resharder, shard_uid, ctx);
-                    },
-                );
+
+            if self
+                .chain_store
+                .get_block_hash_by_height(event.resharding_block.height)
+                .is_ok_and(|hash| hash == event.resharding_block.hash)
```

Comment:
mini nit: maybe split it into two separate conditions with early `continue`?

Reply:
Made the changes but I liked the original if check more